src/core/bpf-firewall.c

```diff
@@ -811,44 +817,35 @@ int bpf_firewall_supported(void) {
          * - whether the unified hierarchy is being used
          * - the BPF implementation in the kernel supports BPF_PROG_TYPE_CGROUP_SKB programs, which we require
          * - the BPF implementation in the kernel supports the BPF_PROG_DETACH call, which we require
+         *
+         * Note that when BPF_FIREWALL_UNSUPPORTED is set, bpf_firewall_unsupported_reason must always be
+         * set too.
          */
         if (supported >= 0)
                 return supported;
 
         r = cg_unified_controller(SYSTEMD_CGROUP_CONTROLLER);
         if (r < 0)
                 return log_error_errno(r, "Can't determine whether the unified hierarchy is used: %m");
-        if (r == 0) {
-                bpf_firewall_unsupported_reason =
-                        log_debug_errno(SYNTHETIC_ERRNO(EUCLEAN),
-                                        "Not running with unified cgroups, BPF firewalling is not supported.");
-                return supported = BPF_FIREWALL_UNSUPPORTED;
-        }
+        if (r == 0)
+                return log_unsupported(SYNTHETIC_ERRNO(EUCLEAN),
+                                       "Not running with unified cgroups, BPF firewalling is not supported.");
```

Comment:
is a macro really the best way to address that? this feels like a function call, but has local side effects (i.e. touches the `supported` variable), which I find a bit icky. Why not do:

```
if (r == 0) {
        r = log_debug_errno(SYNTHETIC_ERRNO(EUCLEAN), "Not running with unified cgroups, BPF firewalling is not supported.");
        goto unsupported;
}
```

or so?

yes a tiny bit longer, but less magic?